### DISCUSSION TITLE
Together.ai implementation

### DISCUSSION THREAD

**Discussion by RyanLucchese** [zed-industries/zed#7213] [Show and tell]
Hi all! Zed is awesome. I'd like to implement a together.ai provider as an alternative to OpenAI for the Assistant.

# Problem Statement
There are many great open source models, including specialized code models, that could benefit Zed users besides GPT. Many users would like to utilize open source models because they can provide some specific capabilities, and they may prefer a different privacy policy than OpenAI. Some users may want to run their own fine-tuned model as well.

# Proposed Solution
Implementing a together.ai provider would be a straight-forward way to access over 100 open source models for the Assistant. Together.ai provides an OpenAI compatible API, which would make integration easy. I'd be happy to submit a PR for the integration.

**Comment by IlyesTeffaha** [comment]
this sounds like a good idea

**Comment by laibe** [comment]
You can already use it by configuring the zed settings.json to use the together.ai as `api_url`, for example:

```json
{
  "assistant": {
    "default_model": {
      "provider": "openai",
      "model": "meta-llama/Meta-Llama-3.1-70B-Instruct-Turbo"
    },
    "version": "2"
  },
  "language_models": {
    "openai": {
      "version": "1",
      "api_url": "https://api.together.xyz/v1",
      "available_models": [
        {
          "name": "meta-llama/Meta-Llama-3.1-70B-Instruct-Turbo",
          "max_tokens": 512
        }
      ]
    }
  }
}
```

**Comment by benjaminbollen** [comment]
how would I set the API key from Together.ai though, any pointers of people who use together.ai in zed.dev?
